### PR TITLE
PWN-6355 - Add part of sell screen logic

### PR DESCRIPTION
**Pull request by davran1001**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-6355

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/common/mvp/BaseMvpFragment.kt [hunk 1]
```diff
@@ -48,6 +48,10 @@ abstract class BaseMvpFragment<V : MvpView, P : MvpPresenter<V>>(
         showErrorDialog(messageRes = messageResId)
     }
 
+    override fun showErrorMessage(message: String) {
+        showErrorDialog(message)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/common/mvp/BaseMvpFragment.kt:52]: we are not using error dialog anymore, we use snackbars

File: app/src/main/java/org/p2p/wallet/moonpay/clientsideapi/MoonpayClientSideApi.kt [hunk 2]
```diff
@@ -45,6 +46,6 @@ interface MoonpayClientSideApi {
         @Path("token") tokenSymbol: String,
         @Query("apiKey") apiKey: String,
         @Query("quoteCurrencyCode") fiatName: String,
-        @Query("baseCurrencyAmount") tokenAmount: Double
+        @Query("baseCurrencyAmount") tokenAmount: BigDecimal
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/moonpay/clientsideapi/MoonpayClientSideApi.kt:49]: when parsing from the server, BigDecimal can be parsed very strangely  
baseCurrencyAmount: 0.15
but Gson will parse it as 0.149999999999999

File: app/src/main/java/org/p2p/wallet/sell/interactor/SellInteractor.kt [hunk 3]
```diff
@@ -46,10 +49,18 @@ class SellInteractor(
...
+    suspend fun getSellQuoteForSol(solAmount: BigDecimal, fiat: MoonpaySellFiatCurrency): MoonpaySellTokenQuote {
         val solToken = homeLocalRepository.getUserTokens().find(Token.Active::isSOL)
         requireNotNull(solToken) { "SOL token is not found for current user, can't sell" }
 
         return sellRepository.getSellQuoteForToken(solToken, solAmount, fiat)
     }
+
+    suspend fun getAllCurrencies() = currencyRepository.getAllCurrencies()
+
+    suspend fun getMoonpaySellFiatCurrency(): MoonpaySellFiatCurrency {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/interactor/SellInteractor.kt:61]: you can return MoonpaySellFiatCurrency directly in repository

File: app/src/main/java/org/p2p/wallet/sell/ui/error/SellErrorFragment.kt [hunk 4]
```diff
@@ -13,28 +13,43 @@ import org.p2p.wallet.utils.popBackStack
...
 
 private const val ARG_ERROR_STATE = "ARG_ERROR_STATE"
+private const val ARG_MIN_AMOUNT = "ARG_MIN_AMOUNT"
 
 class SellErrorFragment : BaseFragment(R.layout.fragment_sell_error) {
 
     companion object {
-        fun create(errorState: SellScreenError): SellErrorFragment =
+        fun create(
+            errorState: SellScreenError,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/error/SellErrorFragment.kt:25]: you can pass minAmount to particular SellScreenError to couple it more

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadFragment.kt [hunk 5]
```diff
@@ -33,6 +37,15 @@ class SellPayloadFragment :
             buttonSend.setOnClickListener {
                 presenter.cashOut()
             }
+            editTextInputToken.setOnTokenAmountChangeListener {
+                presenter.onTokenAmountChanged(it)
+            }
+            editTextAmount.setOnCurrencyAmountChangeListener {
+                presenter.onCurrencyAmountChanged(it)
+            }
+            textViewSellAmount.setOnClickListener {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadFragment.kt:46]: textViewAvailableAmount

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadFragment.kt [hunk 6]
```diff
@@ -33,6 +37,15 @@ class SellPayloadFragment :
             buttonSend.setOnClickListener {
                 presenter.cashOut()
             }
+            editTextInputToken.setOnTokenAmountChangeListener {
+                presenter.onTokenAmountChanged(it)
+            }
+            editTextAmount.setOnCurrencyAmountChangeListener {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadFragment.kt:43]: editTextFiatAmount

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadFragment.kt [hunk 7]
```diff
@@ -45,12 +58,65 @@ class SellPayloadFragment :
...
+            SellErrorFragment.create(
+                errorState = SellErrorFragment.SellScreenError.NOT_ENOUGH_AMOUNT,
+                minAmount = minAmount
+            )
+        )
     }
 
-    override fun setMinSolToSell(minAmount: BigDecimal, tokenSymbol: String) {
+    override fun updateValues(
+        quoteAmount: String,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadFragment.kt:79]: can be extracted to a State class and passed here

File: app/src/main/java/org/p2p/wallet/common/mvp/BaseMvpActivity.kt [hunk 8]
```diff
@@ -37,6 +37,9 @@ abstract class BaseMvpActivity<V : MvpView, P : MvpPresenter<V>> : AppCompatActi
         showErrorDialog(e)
     }
 
+    override fun showErrorMessage(message: String) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/common/mvp/BaseMvpActivity.kt:40]: again, no need for error message

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadFragment.kt [hunk 9]
```diff
@@ -55,17 +72,77 @@ class SellPayloadFragment :
...
+            )
+        )
+    }
+
+    override fun updateViewState(newState: ViewState) = with(binding) {
+        editTextFiatAmount.setText(newState.quoteAmount)
+        editTextFiatAmount.setHint(newState.fiatSymbol)
+        textViewFee.text = getString(R.string.sell_included_fee, newState.fee)
+        textViewRate.text = getString(R.string.sell_sol_fiat_value, newState.fiat)
+        binding.editTextTokenAmount.setHint(newState.tokenSymbol)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadFragment.kt:97]: already have binding in scope

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadFragment.kt [hunk 10]
```diff
@@ -55,17 +72,77 @@ class SellPayloadFragment :
...
 }
+
+data class CashOutButtonState(
+    val isEnabled: Boolean,
+    @ColorRes val backgroundColor: Int,
+    @ColorRes val textColor: Int,
+    val text: String
+)
+
+data class ViewState(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadFragment.kt:140]: extract to contract / or name with context such as SellPayloadFragmentViewState

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt [hunk 11]
```diff
@@ -12,48 +24,48 @@ import org.p2p.wallet.user.interactor.UserInteractor
...
+                view?.showLoading(isVisible = true)
                 checkForSellLock()
-                initView()
+                userSolToken = userInteractor.getUserSolToken()
+                loadCurrencies()
+                checkForMinAmount()
+                startLoadSellQuoteJob()
             } catch (e: Throwable) {
-                Timber.e("Error on init view $e")
+                Timber.e("Error on loading data from Moonpay $e")
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt:59]:
```suggestion
                Timber.e(e,"Error on loading data from Moonpay")
```

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt [hunk 12]
```diff
@@ -65,21 +77,57 @@ class SellPayloadPresenter(
...
+                        fiat = fiat,
+                        minSolToSell = userSelectedAmount.toString(),
+                        tokenSymbol = Constants.SOL_SYMBOL,
+                        fiatSymbol = selectedFiat.currencySymbol,
+                        userBalance = userSolToken?.total.orZero().toString()
+                    )
+                    view?.updateViewState(viewState)
+                    delay(SELL_QUOTE_REQUEST_DEBOUNCE_TIME)
+                } catch (e: CancellationException) {
+                    Timber.e(e)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt:124]:
```suggestion
                    Timber.i(e)
```

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt [hunk 13]
```diff
@@ -88,9 +136,61 @@ class SellPayloadPresenter(
...
+                )
+            }
+        }
+        view?.reset().takeIf { !buttonState.isEnabled }
+        view?.setButtonState(buttonState)
+    }
+
+    override fun onCurrencyAmountChanged(newValue: String) {
+        val newCurrencyAmount = newValue.toBigDecimalOrZero()
+        val newTokenAmount = newCurrencyAmount.divide(tokenPrice, 2, RoundingMode.HALF_UP)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt:190]: 2 to const

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt [hunk 14]
```diff
@@ -88,9 +136,61 @@ class SellPayloadPresenter(
...
+                startLoadSellQuoteJob()
+                CashOutButtonState(
+                    isEnabled = true,
+                    backgroundColor = R.color.bg_night,
+                    textColor = R.color.text_snow,
+                    text = resourceProvider.getString(R.string.common_cash_out)
+                )
+            }
+        }
+        view?.reset().takeIf { !buttonState.isEnabled }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt:184]: reset would be called anyways, take if doesn't work such way

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt [hunk 15]
```diff
@@ -88,9 +136,61 @@ class SellPayloadPresenter(
...
-            fiatSymbol = "usd",
-            tokenAmountToSell = "0.005",
+            fiatSymbol = fiat?.symbol.orEmpty(),
+            tokenAmountToSell = userSelectedAmount.toString(),
         )
         view?.showMoonpayWidget(url = moonpayUrl)
     }
+
+    override fun onTokenAmountChanged(newValue: String) {
+        val newDoubleValue = newValue.toBigDecimalOrZero()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt:146]: newDoubleValue but type is BigDecimal

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt [hunk 16]
```diff
@@ -12,48 +24,48 @@ import org.p2p.wallet.user.interactor.UserInteractor
...
     private val tokenKeyProvider: TokenKeyProvider,
-    private val moonpayWidgetUrlBuilder: MoonpayWidgetUrlBuilder
+    private val moonpayWidgetUrlBuilder: MoonpayWidgetUrlBuilder,
+    private val userInteractor: UserInteractor,
+    private val resourceProvider: ResourcesProvider
 ) : BasePresenter<SellPayloadContract.View>(),
     SellPayloadContract.Presenter {
 
-    private sealed interface ViewState {
-        object Initial : ViewState
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt:29]: how do you handle returning back from the Moonpay without finishing flow? The inputs still there?

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt [hunk 17]
```diff
@@ -12,48 +24,48 @@ import org.p2p.wallet.user.interactor.UserInteractor
...
+                view.showErrorScreen()
             } finally {
                 view.showLoading(isVisible = false)
             }
         }
     }
 
     private suspend fun checkForSellLock() {
-        val userTransactionInProcess = getUserTransactionInProcess()
+        val userTransactionInProcess = isUserHasTransactionInProcess()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt:68]: why rename?

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt (hunk 14 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt]: yeah fixed it

File: app/src/main/java/org/p2p/wallet/sell/ui/error/SellErrorFragment.kt (hunk 4 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/sell/ui/error/SellErrorFragment.kt:25]: I think, we need here better abstraction, cause there is no way move minAmount to enum class, it was predefined with default values )

File: app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt (hunk 16 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/sell/ui/payload/SellPayloadPresenter.kt:29]: hmm I don't understand what you mean

**Pull request merged**